Src/WindowsServer/WindowsServer.Shared/Implementation/AzureMetadataRequestor.cs

```diff
@@ -0,0 +1,115 @@
+﻿namespace Microsoft.ApplicationInsights.WindowsServer.Implementation
+{
+    using System;
+    using System.Collections.Generic;
+    using System.IO;
+    using System.Net;
+    using System.Net.Http;
+    using System.Runtime.Serialization.Json;
+    using System.Threading.Tasks;
+#if NETSTANDARD1_3
+    using System.Net.Http;
+#endif
+    using Extensibility;
+    using Microsoft.ApplicationInsights.WindowsServer.Implementation.DataContracts;
+
+    internal class AzureMetadataRequestor : IAzureMetadataRequestor
+    {
+        /// <summary>
+        /// Azure Instance Metadata Service exists on a single non-routable IP on machines configured
+        /// by the Azure Resource Manager. See <a href="https://go.microsoft.com/fwlink/?linkid=864683">to learn more.</a>
+        /// </summary>
+        internal const string AzureImsApiVersion = "api-version=2017-08-01"; // this version has the format=text capability
+        internal const string AzureImsJsonFormat = "format=json";
+        internal const int AzureImsMaxResponseBufferSize = 512;
+
+        /// <summary>
+        /// Default timeout for the web requests made to obtain Azure IMS data. Internal to expose to tests.
+        /// </summary>
+        internal TimeSpan AzureImsRequestTimeout = TimeSpan.FromSeconds(10);
+
+        /// <summary>
+        /// Private function for mocking out the actual call to IMS in unit tests. Available to internal only.
+        /// </summary>
+        /// parameter sent to the func is a string representing the Uri to request Azure IMS data from.
+        /// <returns>An instance of AzureInstanceComputeMetadata or null.</returns>
+        private Func<string, Task<AzureInstanceComputeMetadata>> azureIMSRequestor = null;
+
+        private string baseImdsUri = "http://169.254.169.254/metadata/instance/compute";
+
+        internal AzureMetadataRequestor(Func<string, Task<AzureInstanceComputeMetadata>> makeAzureIMSRequestor = null)
+        {
+            this.azureIMSRequestor = makeAzureIMSRequestor;
+        }
+
+        /// <summary>
+        /// Gets or sets the base URI for the Azure Instance Metadata service. Internal to allow overriding in test.
+        /// </summary>
+        internal string BaseAimsUri
+        {
+            get => this.baseImdsUri;
+            set => this.baseImdsUri = value;
+        }
+
+        public async Task<AzureInstanceComputeMetadata> GetAzureComputeMetadata()
```

Comment:
Couple of comments.
1. The usual naming convention is for async (task-returning) methods to have the suffix `Async`. i.e. `GetAzureComputeMetadataAsync`.
2. When you have an async method that is simply a straight-line call to another async method, you can (and should) remove the "async" and "await" because that adds unnecessary (compiler-generated) overhead. i.e.
```C#
public Task<AzureInstanceComputeMetadata> GetAzureComputeMetadataAsync()
{
    string metadataRequestUri = $"etc.";
    return this.MakeAzureMetadataRequest(metadataRequestUrl);
}
```

Reply:
Convention is always good to follow, makes it trivial to follow along. Thanks for calling this out.

To the compiler overhead issue, I will try and keep chained async calls out of my future PRs. However, in this case I'm going from a public API call `GetAzureComputeMetadataAsync()` all the way down to the call to `HttpClient`'s `.GetStreamAsync`. I realize I've placed a few interim methods in between the two (to try and keep the meaning of each method as concise as possible - potentially _too_ concise), but it's all about that `.GetStreamAsync` here.